Source/WTF/wtf/text/WYHasher.h

```diff
@@ -0,0 +1,264 @@
+#include <wtf/Vector.h>
+#include <wtf/text/LChar.h>
+
+namespace WTF {
+
+class WYHasher {
+    WTF_MAKE_FAST_ALLOCATED;
+public:
+    static constexpr const unsigned flagCount = 8;
+    static constexpr const unsigned maskHash = (1U << (sizeof(unsigned) * 8 - flagCount)) - 1;
+    static constexpr const uint64_t secret[4] = { 0xa0761d6478bd642full, 0xe7037ed1a0b428dbull, 0x8ebc6af09c88c6e3ull, 0x589965cc75374cc3ull };
+
+    WYHasher() = default;
+
+    template<typename T, typename Converter>
+    static unsigned computeHashAndMaskTop8Bits(const T* data, unsigned characterCount)
+    {
+        return finalizeAndMaskTop8Bits(computeHashImpl<T, Converter>(data, characterCount));
+    }
+
+    template<typename T>
+    static constexpr unsigned computeHashAndMaskTop8Bits(const T* data, unsigned characterCount)
+    {
+        return finalizeAndMaskTop8Bits(computeHashImpl<T>(data, characterCount));
+    }
+
+    template<typename T, unsigned charactersCount>
+    static constexpr unsigned computeLiteralHashAndMaskTop8Bits(const T (&characters)[charactersCount])
+    {
+        return computeHashAndMaskTop8Bits<T>(characters, charactersCount - 1);
+    }
+
+private:
+    static constexpr unsigned finalizeAndMaskTop8Bits(unsigned hash)
+    {
+        return avoidZero(hash & WYHasher::maskHash);
+    }
+
+    ALWAYS_INLINE static constexpr unsigned avoidZero(unsigned hash)
+    {
+        if (hash)
+            return hash;
+        return 0x80000000 >> flagCount;
+    }
+
+    ALWAYS_INLINE static constexpr uint64_t wyrot(uint64_t x)
+    {
+        return (x >> 32) | (x << 32);
+    }
+
+    ALWAYS_INLINE static constexpr void wymum(uint64_t* A, uint64_t* B)
+    {
+#if defined(__SIZEOF_INT128__)
+        __uint128_t r = *A;
+        r *= *B;
+        *A = static_cast<uint64_t>(r);
+        *B = static_cast<uint64_t>(r >> 64);
+#else
+        uint64_t ha = *A >> 32;
+        uint64_t hb = *B >> 32;
+        uint64_t la = static_cast<uint32_t>(*A);
+        uint64_t lb = static_cast<uint32_t>(*B);
+        uint64_t hi;
+        uint64_t lo;
+        uint64_t rh = ha * hb;
+        uint64_t rm0 = ha * lb;
+        uint64_t rm1 = hb * la;
+        uint64_t rl = la * lb;
+        uint64_t t = rl + (rm0 << 32);
+        uint64_t c = t < rl;
+
+        lo = t + (rm1 << 32);
+        c += lo < t;
+        hi = rh + (rm0 >> 32) + (rm1 >> 32) + c;
+        *A = lo;
+        *B = hi;
+#endif
+    }
+
+    ALWAYS_INLINE static constexpr uint64_t wymix(uint64_t A, uint64_t B)
+    {
+        wymum(&A, &B);
+        return A ^ B;
+    }
+
+#define enable_is_constant_evaluated 0
+
+    template<typename T>
+    struct Reader16Bit {
+        ALWAYS_INLINE static constexpr uint64_t wyr2(const T* p)
+        {
+            ASSERT(sizeof(T) == 2);
+            uint64_t firstByte = static_cast<uint8_t>(p[0]);
+            uint64_t secondByte = static_cast<uint8_t>(p[0] >> 8);
+            return (firstByte << 16) | (secondByte << 8) | secondByte;
```

Comment:
This looks strange, why?
I think it should be `return unalignedLoad<uint16_t>(p);`.

Reply:
Discussed offline. This is not `wyr2`, it is `wyr3`. So, we should rename this function to avoid confusion.

Reply:
Here `wyr2` is equivalent to `wyr3` in https://github.com/wangyi-fudan/wyhash/blob/77e50f267fbc7b8e2d09f2d455219adb70ad4749/wyhash.h#L116C3-L116C3, which is used for both `BIG_ENDIAN` and `LITTLE_ENDIAN`.  So, just want to be consistent with original algorithm. I will rename `wry2` to `wyr3`.